internal/controller/stages/event_handlers.go

```diff
@@ -317,61 +343,71 @@ func (u *updatedArgoCDAppHandler[T]) Generic(
 }
 
 // Update implements TypedEventHandler.
-func (u *updatedArgoCDAppHandler[T]) Update(
+func (u *stageEnqueuerForArgoCDChanges[T]) Update(
 	ctx context.Context,
 	e event.TypedUpdateEvent[T],
 	wq workqueue.TypedRateLimitingInterface[reconcile.Request],
 ) {
-	if appHealthOrSyncStatusChanged(ctx, e) {
-		newApp := any(e.ObjectNew).(*argocd.Application) // nolint: forcetypeassert
+	if !appHealthOrSyncStatusChanged(ctx, e) {
+		return
+	}
 
-		stageRef, ok := newApp.Annotations[kargoapi.AnnotationKeyAuthorizedStage]
-		if !ok {
-			return
-		}
-		parts := strings.SplitN(stageRef, ":", 2)
-		if len(parts) != 2 {
-			return
-		}
-		projectName, stageName := parts[0], parts[1]
+	newApp, ok := any(e.ObjectNew).(*argocd.Application)
+	if !ok {
+		return
+	}
 
-		logger := logging.LoggerFromContext(ctx)
-		stage := &kargoapi.Stage{}
-		if err := u.kargoClient.Get(
-			ctx,
-			types.NamespacedName{
-				Namespace: projectName,
-				Name:      stageName,
-			},
-			stage,
-		); err != nil {
-			if client.IgnoreNotFound(err) != nil {
-				logger.Error(
-					err,
-					"error getting Stage for Application",
-					"namespace", projectName,
-					"stage", stageName,
-					"app", newApp.Name,
-				)
-			}
-			return
+	stageRef, ok := newApp.Annotations[kargoapi.AnnotationKeyAuthorizedStage]
+	if !ok {
+		return
+	}
+	parts := strings.SplitN(stageRef, ":", 2)
+	if len(parts) != 2 {
+		return
+	}
+	projectName, stageName := parts[0], parts[1]
+
+	logger := logging.LoggerFromContext(ctx)
+	stage := &kargoapi.Stage{}
+	if err := u.kargoClient.Get(
+		ctx,
+		types.NamespacedName{
+			Namespace: projectName,
+			Name:      stageName,
+		},
+		stage,
+	); err != nil {
+		if client.IgnoreNotFound(err) != nil {
+			logger.Error(
+				err,
+				"error getting Stage for Application",
+				"namespace", projectName,
+				"stage", stageName,
+				"app", newApp.Name,
+			)
 		}
+		return
+	}
 
-		wq.Add(
-			reconcile.Request{
-				NamespacedName: types.NamespacedName{
-					Namespace: stage.Namespace,
-					Name:      stage.Name,
-				},
-			},
-		)
-		logger.Debug(
-			"enqueued Stage for reconciliation",
-			"namespace", stage.Namespace,
-			"stage", stage.Name,
-			"app", newApp.Name,
-		)
+	// If the Stage is a control flow Stage, there is no need to reconcile it.
+	if stage.IsControlFlow() {
+		return
```

Comment:
Previously, this would always result in a reconciliation — even if there wasn't any actual interest.